Automattic-Tracks-iOSTests/CrashLoggingTests.swift

```diff
@@ -0,0 +1,141 @@
+import XCTest
+@testable import AutomatticTracks
+
+class CrashLoggingTests: XCTestCase {
+
+    private let crashLogging = CrashLogging()
+    private var mockDataProvider = MockCrashLoggingDataProvider()
+
+    override func setUp() {
+        mockDataProvider.sentryDSN = validDSN
+    }
+
+    override func tearDown() {
+        mockDataProvider.reset()
+    }
+
+    func testInitializationWithInvalidDSN() {
+        let expectation = XCTestExpectation(description: "Intialization should fail with a DSN error")
+
+        mockDataProvider.sentryDSN = invalidDSN
+        mockDataProvider.didLogErrorCallback = { event in
+            if event.message == "Project ID path component of DSN is missing" {
+                expectation.fulfill()
+            }
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testInitializationWithValidDSNFormat() {
+        let expectation = XCTestExpectation(description: "Intialization should pass")
+        expectation.isInverted = true
+
+        mockDataProvider.didLogErrorCallback = { event in
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testThatUserDataIsBeingStoredForUseInCrashLogs() {
+        mockDataProvider._currentUser = testUser
+        CrashLogging.start(withDataProvider: mockDataProvider)
+
+        XCTAssert(CrashLogging.sharedInstance.currentUser.email == testUser.email)
+        XCTAssert(CrashLogging.sharedInstance.cachedUser?.email == testUser.email)
+    }
+
+    func testThatEventsAreNotSentWhenUserOptsOut() {
+        mockDataProvider.userHasOptedOut = true
+
+        let expectation = XCTestExpectation(description: "Events should not be sent if user has opted out")
+
+        CrashLogging.sharedInstance.shouldSendEventCallback = { shouldSendEvent in
+            expectation.isInverted = shouldSendEvent   // fail if `shouldSendEvent` is true
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+        CrashLogging.logMessage("This is a test")
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+    func testThatEventsAreSentWhenUserOptsIn() {
+        mockDataProvider._currentUser = testUser
+
+        let expectation = XCTestExpectation(description: "Events be sent if user has opted in")
+
+        CrashLogging.sharedInstance.shouldSendEventCallback = { shouldSendEvent in
+            expectation.isInverted = !shouldSendEvent   // succeed if `shouldSendEvent` is true
+            expectation.fulfill()
+        }
+
+        CrashLogging.start(withDataProvider: mockDataProvider)
+        CrashLogging.logMessage("This is a test")
+
+        wait(for: [expectation], timeout: 1)
+    }
+
+//    func testThatLoggedErrorsContainAStackTrace() {
```

Comment:
❓ What is the plan for these commented out tests?

Reply:
I'm not 100% sure why these fail – I'd like to reach out to Sentry to figure that out, but these are very much nice-to-haves, rather than must-haves. I suspect it's something to do with the release configuration and/or they don't run on the simulator, but I'd like to get them working at some point for full code coverage :)

Reply:
Cool. Perhaps a comment in the code would be nice so someone will not delete them. I know I would. 😄

Reply:
Agreed – added in 3f8da24a122477cc98791574e41e000e35dcfb6e